### PR TITLE
Add DeepSpeed 103B GPT pretraining benchmark and standardize containers for B200

### PR DESCRIPTION
**Pull request by paragao**
## Summary

- Add a **103B-parameter GPT pretraining benchmark** using Megatron-DeepSpeed on NVIDIA B200 clusters with 3D parallelism (TP/PP/DP) and DeepSpeed ZeRO optimization
- **Rewrite the DeepSpeed README** as a use-case-focused guide covering all three test cases (GPT-103B pretraining, QLoRA fine-tuning, Llama2 fine-tuning) with best-practice configuration recommendations
- **Standardize the QLoRA Dockerfile** with the same infrastructure best practices as the pretraining container

### New files

| File | Description |
|------|-------------|
| `pretrain_gpt_103b.sbatch` | Parameterized Slurm script for 103B GPT pretraining (TP, PP, ZeRO, fusions, seq length all configurable via env vars) |
| `parse_results.py` | Parses Megatron-DeepSpeed training logs into benchmark JSON |
| `configs/ds_config_103b_template.json` | Reference DeepSpeed config for the 103B model |

### Modified files

| File | Change |
|------|--------|
| `0.deepspeed.dockerfile` | Rewritten: `pytorch:25.04-py3` base (CUDA 12.9), EFA 1.47.0, NCCL 2.29.3, GDRCopy v2.5.1, OFI-NCCL symlinks, `ld.so.conf.d` library discovery |
| `1.build-image.sbatch` | Added `mkdir -p /fsx/apps`, fixed job name |
| `Makefile` | Rewritten: `train` target uses best config (TP=8, PP=8, fusions), `parse` target for log parsing, `help` with descriptions |
| `README.md` | Rewritten as use-case index with prerequisites, container setup, data prep, running instructions, best practices, and known issues |
| `qlora/Dockerfile` | Standardized: same `pytorch:25.04-py3` base, EFA 1.47, NCCL 2.29.3, GDRCopy 2.5.1, proper NCCL/EFA env vars, `expandable_segments:True` |
| `qlora/requirements.txt` | Updated comment to reflect new CUDA 12.9 index URL |

### Key best practices documented

- Maximize pipeline parallelism (TP=8, PP=8) with kernel fusions enabled for best throughput
- ZeRO-0 over ZeRO-1 when data-parallel group size is small
- ZeRO-2/3 require `--no-pipeline-parallel` (Megatron-DeepSpeed PipelineEngine limitation)
- Do not set `NCCL_ALGO=Tree` on EFA clusters (causes hangs)
- Use `expandable_segments:True` (capital T) in pytorch:25.04 containers
- Use `python3 -m torch.distributed.run` instead of `torchrun` (shebang compatibility)

## Testing

All configurations were validated on an 8-node B200 HyperPod cluster (64 GPUs total) with 50 training steps each.

### CODE REVIEW

File: 3.test_cases/pytorch/deepspeed/0.deepspeed.dockerfile [hunk 1]
```diff
@@ -55,69 +56,99 @@ RUN rm -rf /root/.ssh/ \
...
+ && chmod a+x ${OPEN_MPI_PATH}/bin/mpirun
+
+# ============================================================
+# 9. Python packages for DeepSpeed training
+# ============================================================
+RUN pip3 install --no-cache-dir \
+    awscli pynvml \
+    transformers==${TRANSFORMERS_VERSION} \
+    sentencepiece python-etcd \
+    deepspeed accelerate
```
Comment (KeitaW) [inline 3.test_cases/pytorch/deepspeed/0.deepspeed.dockerfile]: ### Unpinned Python packages

`deepspeed` and `accelerate` are installed without version pins. Per repo conventions, dependencies should have at least an upper-bound pin to prevent silent breakage from major version changes. For example:

```suggestion
RUN pip3 install --no-cache-dir \
    awscli pynvml \
    transformers==${TRANSFORMERS_VERSION} \
    sentencepiece python-etcd \
    deepspeed>=0.16,<1.0 accelerate>=1.0,<2.0
```

I'd also double-check the QLoRA `requirements.txt` for the same issue — any packages there without upper bounds should get them too.

File: 3.test_cases/pytorch/deepspeed/gpt/parse_results.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
+#!/usr/bin/env python3
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+"""
+parse_results.py - Parse Megatron-DeepSpeed training logs into benchmark JSON.
+
+Reads Slurm log files, extracts per-step metrics, and produces JSON files
+matching the existing benchmark-results schema at:
+  s3://paragao-new-nemo-squash-container/benchmark-results/b200/
+
```
Comment (KeitaW) [inline 3.test_cases/pytorch/deepspeed/gpt/parse_results.py:10]: ### Personal S3 bucket reference

This looks like a personal bucket path in the module docstring. Since this will be visible in the public repo, I'd suggest either removing it or replacing it with a generic path.

```suggestion
matching the benchmark-results schema.

Usage:
```

File: 3.test_cases/pytorch/deepspeed/1.build-image.sbatch [hunk 3]
```diff
@@ -14,6 +14,10 @@ set -euxo pipefail
 : "${APPS_PATH:=/fsx/apps}"
 : "${IMAGE:=$APPS_PATH/deepspeed.sqsh}"
 
+# Ensure output directory exists
+mkdir -p ${APPS_PATH}
+mkdir -p logs
```
Comment (KeitaW) [inline 3.test_cases/pytorch/deepspeed/1.build-image.sbatch]: ### Unquoted variable expansions

The repo convention calls for quoting all variable expansions to guard against word splitting. While these are unlikely to contain spaces in practice, it's good hygiene and consistent with the rest of the codebase.

```suggestion
mkdir -p "${APPS_PATH}"
mkdir -p logs
```

The same applies throughout `pretrain_gpt_103b.sbatch` — several `${IMAGE}`, `${HOSTFILE}`, and other expansions are unquoted, especially in the `srun` arguments and the `bash -c` string interpolation at the bottom of the script.

File: 3.test_cases/pytorch/deepspeed/pretrain_gpt_103b.sbatch [hunk 4]
```diff
@@ -0,0 +1,280 @@
+#!/bin/bash
+
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+#SBATCH --exclusive
+#SBATCH --job-name=deepspeed-pretrain-103b
+#SBATCH --output=logs/%x_%j.out
+#SBATCH --error=logs/%x_%j.err
+#SBATCH --partition=b200
```
Comment (KeitaW) [inline 3.test_cases/pytorch/deepspeed/pretrain_gpt_103b.sbatch]: ### Hardcoded partition name

The `b200` partition is specific to your cluster. Other users won't have this partition, and the `#SBATCH` directive takes precedence over `sbatch --partition=...` command-line overrides. Since the Makefile already passes `--partition=$(PARTITION)` at submission time, I'd suggest removing this directive entirely and letting users specify the partition when submitting. That way `make train PARTITION=my-partition` works without the `#SBATCH` line overriding it.

Alternatively, add a comment noting this must be changed for other clusters.

**Review by KeitaW — COMMENTED** [review]
## Review Batch 2/4 — Deployment Pipeline

File: 3.test_cases/pytorch/deepspeed/pretrain_gpt_103b.sbatch [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+export NUM_GPUS_PER_NODE=8
+
+# ============================================================
+# Network settings for EFA + NCCL
+# ============================================================
+export FI_LOG_LEVEL=1
+export FI_PROVIDER=efa
+export FI_EFA_USE_HUGE_PAGE=0
+export NCCL_SOCKET_IFNAME=^docker,lo,veth
+export NCCL_P2P_NET_CHUNKSIZE=2048576
```
Comment (KeitaW) [inline 3.test_cases/pytorch/deepspeed/pretrain_gpt_103b.sbatch]: ### NCCL_P2P_NET_CHUNKSIZE value looks off

This value is `2048576` (2,048,576). If 2 MB was intended, the conventional power-of-two value would be `2097152` (2×1024×1024). If it's an intentional non-power-of-two tuning choice that works well in your benchmarks, a brief comment explaining why would be helpful.

```suggestion
export NCCL_P2P_NET_CHUNKSIZE=2097152
```

File: 3.test_cases/pytorch/deepspeed/README.md [hunk 6]
```diff
@@ -1,87 +1,179 @@
...
+    --export=ALL,TP=8,PP=4,ZERO_STAGE=1,MICRO_BATCH_SIZE=2,CONFIG_NAME=my_config \
+    pretrain_gpt_103b.sbatch
+```
+
+#### Environment variables
+
+| Variable | Default | Description |
+|----------|---------|-------------|
+| `TP` | 8 | Tensor parallel size |
+| `PP` | 2 | Pipeline parallel size |
```
Comment (KeitaW) [inline 3.test_cases/pytorch/deepspeed/README.md]: ### Default PP doesn't match recommended config

The table shows `PP` defaults to `2`, but the recommended "best config" (in the Makefile and the `make train` example above) uses `PP=8`. Users running the sbatch directly without `make train` will get `PP=2`. Consider either changing the default to `8` or adding a note here that these are safe baselines and `make train` uses the optimized settings.

```suggestion
| `PP` | 2 | Pipeline parallel size (use PP=8 for best throughput, see `make train`) |
```

File: 3.test_cases/pytorch/deepspeed/0.deepspeed.dockerfile (hunk 1 above)
Comment (paragao) [reply 3.test_cases/pytorch/deepspeed/0.deepspeed.dockerfile]: Fixed. Pinned `deepspeed>=0.16,<1.0` and `accelerate>=1.0,<2.0` in the Dockerfile. QLoRA `requirements.txt` already had proper upper bounds — no changes needed there.

File: 3.test_cases/pytorch/deepspeed/gpt/parse_results.py (hunk 2 above)
Comment (paragao) [reply 3.test_cases/pytorch/deepspeed/gpt/parse_results.py:10]: Fixed. Replaced with generic placeholder `s3://<YOUR_BUCKET>/benchmark-results/<instance_type>/`. Also parameterized `cluster` and `instance_type` in `build_result_json()` — they now accept CLI args (`--cluster`, `--instance-type`) and fall back to `$CLUSTER_NAME`/`$INSTANCE_TYPE` env vars.

File: 3.test_cases/pytorch/deepspeed/1.build-image.sbatch (hunk 3 above)
Comment (paragao) [reply 3.test_cases/pytorch/deepspeed/1.build-image.sbatch]: Fixed. Quoted all variable expansions in `1.build-image.sbatch` and `pretrain_gpt_103b.sbatch` (`$SLURM_JOB_NODELIST`, `${HOSTFILE}`, `${IMAGE}`, `${APPS_PATH}`, `${ENROOT_IMAGE}`). For the `srun bash -c` line, variables are intentionally expanded on the host side before passing to the container — added a comment explaining this.

File: 3.test_cases/pytorch/deepspeed/pretrain_gpt_103b.sbatch (hunk 4 above)
Comment (paragao) [reply 3.test_cases/pytorch/deepspeed/pretrain_gpt_103b.sbatch]: Fixed. Removed the `#SBATCH --partition=b200` directive entirely. The partition is now passed exclusively at submit time via `--partition=` in the Makefile. Also changed the Makefile default partition from `b200` to `dev`.

File: 3.test_cases/pytorch/deepspeed/pretrain_gpt_103b.sbatch (hunk 5 above)
Comment (paragao) [reply 3.test_cases/pytorch/deepspeed/pretrain_gpt_103b.sbatch]: Fixed. Changed from `2048576` to `2097152` (2 x 1024 x 1024). The original value was a typo, not an intentional tuning choice.

File: 3.test_cases/pytorch/deepspeed/README.md (hunk 6 above)
Comment (paragao) [reply 3.test_cases/pytorch/deepspeed/README.md]: Fixed. Updated the README table row to: `| PP | 2 | Pipeline parallel size (best throughput with PP=8, see make train) |`

**Comment by paragao** [comment]
## Addressed all review feedback

**Directory restructure (Batch 1/4):** Moved GPT-103B files into `gpt/` subdirectory with `slurm/` and `configs/` sub-dirs to match repo conventions:

```
gpt/
├── configs/ds_config_103b_template.json
├── slurm/pretrain_gpt_103b.sbatch
└── parse_results.py
```

Updated all path references in Makefile and README.

**Documentation (Batch 4/4):** Added trailing newline to README.md. Updated PP row with note about `make train`.

**Also fixed across these commits:**
- Removed personal S3 bucket references, parameterized cluster/instance metadata
- Removed hardcoded `--partition=b200`, changed default to `dev`
- Pinned `deepspeed>=0.16,<1.0` and `accelerate>=1.0,<2.0` in Dockerfile
- Quoted all variable expansions in build and training scripts
- Corrected `NCCL_P2P_NET_CHUNKSIZE` from `2048576` to `2097152` (power-of-two)

See individual inline replies for details on each item.

File: 3.test_cases/pytorch/deepspeed/README.md [hunk 7]
```diff
@@ -1,87 +1,179 @@
...
+           f.write('\n')
+   "
    ```
 
-3. Convert the Docker image to a squash file with the command below.
+3. Clone Megatron-DeepSpeed and preprocess:
 
    ```bash
-   enroot import -o ${ENROOT_IMAGE} dockerd://deepspeed:latest
+   git clone https://github.com/microsoft/Megatron-DeepSpeed /fsx/deepspeed/Megatron-DeepSpeed
```
Comment (KeitaW) [inline 3.test_cases/pytorch/deepspeed/README.md:73]: ### Megatron-DeepSpeed clone is not pinned to a version

The previous Dockerfile had `ARG MEGATRON_LM_VERSION=core_r0.8.0`, but this was removed. Per the repo convention, external dependencies must be pinned to a version/tag/commit — never HEAD. I'd suggest pinning to whatever commit or tag was validated during testing on the B200 cluster.

```suggestion
   git clone --branch <validated_tag> --depth 1 https://github.com/microsoft/Megatron-DeepSpeed /fsx/deepspeed/Megatron-DeepSpeed
```

File: 3.test_cases/pytorch/deepspeed/1.build-image.sbatch [hunk 8]
```diff
@@ -14,11 +14,15 @@ set -euxo pipefail
...
-if [ -f ${ENROOT_IMAGE}.sqsh ] ; then
-    rm ${ENROOT_IMAGE}.sqsh
+if [ -f "${ENROOT_IMAGE}.sqsh" ] ; then
+    rm "${ENROOT_IMAGE}.sqsh"
 fi
-enroot import -o ${ENROOT_IMAGE}.sqsh dockerd://${ENROOT_IMAGE}:latest
-mv ${ENROOT_IMAGE}.sqsh ${IMAGE}
\ No newline at end of file
+enroot import -o "${ENROOT_IMAGE}.sqsh" "dockerd://${ENROOT_IMAGE}:latest"
+mv "${ENROOT_IMAGE}.sqsh" "${IMAGE}"
```
Comment (KeitaW) [inline 3.test_cases/pytorch/deepspeed/1.build-image.sbatch:28]: ### Missing final newline

The diff shows `\ No newline at end of file` here. Per `.editorconfig` (`insert_final_newline = true`), this should end with a trailing newline.

File: 3.test_cases/pytorch/deepspeed/qlora/Dockerfile [hunk 9]
```diff
@@ -1,61 +1,154 @@
...
 # Copy requirements first for better caching
 COPY requirements.txt .
 
-# Install PyTorch with CUDA 12.8 support
+# Install PyTorch with CUDA 12.9 support
 # Note: torch 2.10+ has a breaking LR scheduler change (strict zip) that is
 # incompatible with some DeepSpeed/transformers versions. Pin to <2.10 until
 # upstream libraries catch up.
-RUN pip install --no-cache-dir 'torch>=2.7.0,<2.10.0' --index-url https://download.pytorch.org/whl/cu128
+RUN pip install --no-cache-dir 'torch>=2.7.0,<2.10.0' --index-url https://download.pytorch.org/whl/cu129
```
Comment (KeitaW) [inline 3.test_cases/pytorch/deepspeed/qlora/Dockerfile:144]: ### Reinstalling PyTorch on top of NGC base image

The base image `nvcr.io/nvidia/pytorch:25.04-py3` already ships with a CUDA 12.9-optimized PyTorch. This line overwrites it with the generic PyPI wheel, which may lack NGC-specific optimizations (cuDNN auto-tuning, NVTX annotations, etc.) and increases image size.

The main Dockerfile (`0.deepspeed.dockerfile`) correctly relies on the base image's PyTorch. I'd suggest either removing this line (the base image's torch satisfies `>=2.7.0,<2.10.0`) or adding a comment explaining why the override is intentional (e.g., a specific version requirement for QLoRA compatibility).
